### ISSUE TITLE
Tracking Issue for XXX

### ISSUE THREAD

**Issue by nikomatsakis** [rust-lang/rust#71668]
This is a tracking issue for the RFC "unsafe blocks in unsafe fn" (rust-lang/rfcs#2585).
The feature gate for the issue is `#![feature(unsafe_block_in_unsafe_fn)]`.

### About tracking issues

Tracking issues are used to record the overall progress of implementation.
They are also uses as hubs connecting to other relevant issues, e.g., bugs or open design questions.
A tracking issue is however *not* meant for large scale discussion, questions, or bug reports about a feature.
Instead, open a dedicated issue for the specific matter and add the relevant feature gate label.

### Steps

- [ ] Introduce an opt-in lint that, when enabled, causes unsafe blocks in unsafe functions to be considered required, and warns if they are absent when an unsafe operation is performed. 
- [ ] Include a suggestion with the lint that can insert required unsafe blocks. This could be as simple as adding a block across the entire function, though more granular insertion is probably better.
- [ ] Adjust documentation to describe the (somewhat unusual) effect of the lint, and to describe the possibility that the lint will be enabled default ([see instructions on rustc-dev-guide][doc-guide])
- [ ] Stabilization PR ([see instructions on rustc-dev-guide][stabilization-guide])
- [ ] Write a blog post describing the change, perhaps?

[stabilization-guide]: https://rustc-dev-guide.rust-lang.org/stabilization_guide.html#stabilization-pr
[doc-guide]: https://rustc-dev-guide.rust-lang.org/stabilization_guide.html#documentation-prs

### Unresolved Questions

* What is the timeline for adding the lint, and cranking up its default level?
* Should the default level depend on the edition?
* Should we ever make this deny-by-default or even a hard error, in a future edition?
* Should we require `cargo fix` to be able to do *something* about this warning before making it even warn-by-default?  And how precise should it be?

### Implementation history

* None so far

**Comment by nikomatsakis** [comment]
cc @RalfJung

**Comment by nikomatsakis** [comment]
A relevant note towards the implementation: 

* Clippy has a [helper function](https://github.com/rust-lang/rust-clippy/blob/9a3b0a05888f7ad002b67bbe61739695967ce929/clippy_lints/src/utils/mod.rs#L1069) for testing whether a lint is enabled.

**Comment by LeSeulArtichaut** [comment]
I'll try to work on the implementation!

**Comment by RalfJung** [comment]
If I understand https://boats.gitlab.io/blog/post/ringbahn/ correctly, that API would also benefit greatly from not treating the body of an `unsafe fn` as `unsafe`: there is a trait with an `unsafe fn`, and

> It’s important to understand what an unsafe methods means: unsafe methods are unsafe to call - meaning that callers must guarantee additional invariants when calling them. However, they are safe to implement (as long as you don’t do anything unsafe inside of them). So when a user implements the Event trait, they are given additional guarantees about how this method will be called that normally they could not assume.

@withoutboats did I understand this correctly?

**Comment by withoutboats** [comment]
@RalfJung I don't think its a particularly new situation. Basically, the guarantee that you are getting from it being unsafe are the guarantees you need to be able to call [these unsafe methods](https://docs.rs/iou/0.2.0/iou/struct.SubmissionQueueEvent.html#method.prep_read_vectored) in the SubmissionQueueEvent that is passed as an argument.

In other words, an `Event` implementation is going to use probably 1 unsafe operation (calling one of these methods), which it can prove is correct based on the invariant that trait requires the caller to maintain. This isn't really different from any other unsafe code.

What I'm trying to say with the blog post is that the Event trait hasn't introduced a new invariant for you to maintain, its introduced a new invariant that *I* maintain so its easier for you to call that unsafe method you need to call to get any work done.

This is just how unsafe trait methods work in general, so its nothing new. But I agree that unsafe trait methods (where the invariant contract is specified separately from the implementer) are a compelling example of where unsafe blocks in unsafe fns seem beneficial.

**Comment by RalfJung** [comment]
> What I'm trying to say with the blog post is that the Event trait hasn't introduced a new invariant for you to maintain, its introduced a new invariant that I maintain so its easier for you to call that unsafe method you need to call to get any work done.

Is there a realistic safe implementation of that method that ignores the extra invariant provided by the caller, or is the expectation that most/all implementations will have to do something `unsafe` justified by that invariant?

**Comment by withoutboats** [comment]
Yea I believe some kinds of events should be safe to prepare because they only send `Copy` data into the kernel and so they don't need to worry about any kind of memory safety invariant.

**Comment by LeSeulArtichaut** [comment]
Step 1 can now be checked since #71862 landed.

**Comment by withoutboats** [comment]
I want to give an experience report. I've been writing a lot of unsafe code recently with ringbahn, and I've been thinking about how this change would have impacted my experience and the correctness of my code. Unfortunately, my thoughts are not optimistic.

My experience makes me think a big downside of this change has been overlooked, and that the upside is not so great. The goals of this change I think would be better served by a separate linting infrastructure.

First, I think this change would dramatically impact user flow when writing new and unfinished unsafe code. During the prototype, experiment, build-out phase of creating new code, I am often unsure where exactly the function boundaries should go, what code will be reusable, what code should be abstracted, what code needs to be inlined, etc. As a result, I am frequently moving code in and out of functions. By marking those functions unsafe as I construct them, the code remains in an unsafe scope right now. But this change would require me to set up new unsafe scopes as I make these sorts of edits, greatly increasing the ceremony. Similar to the arguments I've made in favor of Ok-wrapping, we should minimize the disruption to the fluidity of creating new functions during this phase of editing, and I see this as a major downside.

Second, so far two memory safety bugs have been discovered in ringbahn, and neither of them would have been prevented by this change. In both cases, invariants were not being upheld by safe code (in both cases, interestingly, because of implicit destructors). The bugs this change is designed to catch seem easy to identify and unlikely to occur: accidentally calling an unsafe function that you didn't mean to call. I am doubtful that this would meaningfully improve the quality of unsafe code or the experience of writing it.

The utility of this change seems to be in the later phase of editing, once the code has settled down and is no longer seeing large refactors. At that point, it is ideal for each unsafe operation to be annotated with a comment justifying why the invariants necessary are being upheld. However, this change to the language is a poor match for actually achieving that goal: even if users are required to put unsafe operations in unsafe blocks, nothing actually makes them justify those unsafe operations, or do anything but wrap entire function bodies in unsafe blocks and call it a day (as I often do during prototyping).

For those situations, what would be more appropriate in my opinion is a linter that requires every statement containing an unsafe operation to be annotated with a comment justifying its correctness. This would automatically enforce what is already becoming a standard style for finished unsafe code. It would not require this change at all to complete.

**Comment by RalfJung** [comment]
> During the prototype, experiment, build-out phase of creating new code, I am often unsure where exactly the function boundaries should go, what code will be reusable, what code should be abstracted, what code needs to be inlined, etc. As a result, I am frequently moving code in and out of functions. By marking those functions unsafe as I construct them, the code remains in an unsafe scope right now. But this change would require me to set up new unsafe scopes as I make these sorts of edits, greatly increasing the ceremony. Similar to the arguments I've made in favor of Ok-wrapping, we should minimize the disruption to the fluidity of creating new functions during this phase of editing, and I see this as a major downside.

I am fully buying your argument with `Ok`-wrapping, but am unconvinced of it in the context of unsafe blocks. I think that is because with `Ok`-wrapping, even having that added implicitly for the final product seems like a great state of affairs to me, whereas for `unsafe` it feels like a prototyping-only kludge.

So maybe a better solution to your problem, instead of keeping `unsafe fn` internally unsafe (with not just the practical but also the conceptual issues that brings) would be the ability to have `unsafe` blocks outside functions, so you could wrap an entire module in `unsafe` and just not care about safety for the time being? That seems to better reflect your prototyping intent.

Another alternative would be a more convenient way to re-gain the ability to mark the entire body of a function as `unsafe`. For sample code, I personally use
```rust
fn main() { unsafe {
} }
```
but rustfmt will turn this into a double-indent. We could adjust rustfmt, or we could adopt @Centril's proposal of defining functions with `=`, which would permit
```rust
fn main() = unsafe {
}
```
which seems like minimal effort even during prototyping, while at the same time not mixing up the two different roles of "unsafe" that `unsafe fn` currently fails to distinguish.

**Comment by phil-opp** [comment]
> Why do you want an interior safe scope? If its to avoid accidentally performing an unsafe op, as I understand it, linters like the one I've described would easily suffice. 

To me the only difference between safe and unsafe functions is that the latter have additional memory safety related requirements for callers. I see no reason to treat the bodies of safe and unsafe functions differently, so it makes sense to me to use the same mechanism for marking unsafe operations.

Of course we could also create a lint that requires an annotation for each unsafe operation, but this should apply to safe and unsafe functions in the same way. If we created such a lint and at some point decided that this lint is enough to mark unsafe operations, I would see no reason why unsafe blocks shouldn't be made optional in bodies of safe functions too.

> To reiterate my previous claim: as we all know, it is not only the unsafe ops that can cause memory safety inside a module with unsafe code; singling out these ops as the unique points at which extra care is needed gives a false sense of security in my mind.

Again, there is no difference between safe and unsafe function bodies to me. Yes, usage of `unsafe` can pollute a whole module, but it is still useful to mark the places where the actual memory safety is caused. Otherwise, we could also make `unsafe` blocks in safe functions optional as soon as there is some unsafe used somewhere in the same module, which does not sound like a good idea to me.

> I note that neither of you responded directly to my second claim that this is not particularly useful for catching bugs.

It's as useful as `unsafe` blocks in safe functions. Yes, the actual bugs might be outside unsafe blocks, but `unsafe` blocks are still useful because they mark the places where additional invariants are required. So even if the bug is somewhere else, you have a place where you can start your debugging, e.g. by adding additional assertions or debugger breakpoints.

**Comment by Aaron1011** [comment]
@withoutboats 

> To reiterate my previous claim: as we all know, it is not only the unsafe ops that can cause memory safety inside a module with unsafe code; singling out these ops as the unique points at which extra care is needed gives a false sense of security in my mind.

That's true - however, the `unsafe` function being called may have additional safety conditions that are unrelated to the rest of the function. When reviewing code, seeing something like:

```rust
unsafe fn foo() {
    unsafe {
        first_fn();
        second_fn();
    }
    third_fn();
}
```

immediately tells me that `third_fn` cannot cause undefined behavior by itself - it can only do so as a result of interaction with other unsafe code (`first_fn()`, `second_fn()`, or something else).

However, if I instead see something like this:

```rust
unsafe fn foo() {
    first_fn();
    second_fn();
    third_fn();
}
```

It's no longer obvious that `third_fn` is not `unsafe` - I now need to check the function definition to see that that it is actually safe (and therefore imposes no additional preconditions).

> what would be more appropriate in my opinion is a linter that requires every statement containing an unsafe operation to be annotated with a comment justifying its correctness.

Would this be an if-and-only-if lint (e.g. it would error if a safe statement has an "unsafe justification" comment)? If not, I think these types of comments could become stale as code is refactored over time. With `unsafe` blocks in `unsafe` functions, moving safe statements outside of `unsafe` blocks will make it visually obvious when comments become out-of-date. Of course, it's still a good idea to have a comment of some kind, but any kind of justification will be in terms of interactions with `unsafe` code in the same module.

**Comment by AaronKutch** [comment]
Is this in nightly yet? It definitely recognizes the `unsafe_block_in_unsafe_fn` feature, but it still treats unsafe functions as if the entire inside is an `unsafe` block and even warns that `unsafe` blocks are unnecessary.

**Comment by RalfJung** [comment]
@AaronKutch we have testcases that say otherwise, so if you have a concrete example, please report it as a new bug (not in this thread, but feel free to Cc me).

**Comment by LeSeulArtichaut** [comment]
@AaronKutch As specified in the RGC, you need to override the lint level for the `unsafe_op_in_unsafe_fn`. Having it `allow`ed just falls back to the previous behavior.

**Comment by AaronKutch** [comment]
Thank you, I have to add both
```
#![feature(unsafe_block_in_unsafe_fn )]
#![deny(unsafe_op_in_unsafe_fn)]
```

**Comment by RalfJung** [comment]
> However, I note that neither of you responded directly to my second claim that this is not particularly useful for catching bugs. 

Fair. I am not sure how to respond to that -- I don't have any data, just a gut feeling and a number of examples where the `unsafe` scope is IMO too big. You presented an experience report saying that keeping the `unsafe` scope small doesn't help, or at least doesn't help you. Neither is statistically significant, but you have much more experience writing real unsafe code than I do, so there's not much I can say to counter this point.

Maybe it is a matter of coding style? Even if module-level unsafe blocks were a thing, I wouldn't use them. I keep my unsafe blocks as small as possible and try to be extremely aware of what exactly are the unsafe operations in my code. Of course that doesn't mean I can ignore the safe code, but it highlights the places where *additional* checks (on top of "maintain module invariants") are required.

> Why do you want an interior safe scope? If its to avoid accidentally performing an unsafe op, as I understand it, linters like the one I've described would easily suffice. To reiterate my previous claim: as we all know, it is not only the unsafe ops that can cause memory safety inside a module with unsafe code; singling out these ops as the unique points at which extra care is needed gives a false sense of security in my mind.

The unsafe ops are still special as they are a place where *externally defined invariants* must be upheld. In contrast, the rest of the code only needs to maintain my own module-local invariants. So when I audit the code, inside an `unsafe` scope, I need to check more things than outside.

Sure, linters can help, but then the workflow for performing unsafe ops become strangely decoherent between normal `fn` and `unsafe fn` -- there is IMO no good reason for them to follow different rules, and yet they do.

I also proposed some ideas once to be able to mark struct fields as unsafe, which would make mutating them an unsafe operation. Then `unsafe` tracking would help even with maintaining module invariants. (This could also affect auto-trait derivation, but that is yet another discussion.) That is an orthogonal feature but could be part of a "vision" for making `unsafe` more helpful. Or maybe it's not actually helpful, I am not sure.

> I get the sense from talking to you that this conceptual issue is a really big motivator for you.

That is certainly true. I consider this an inconsistency/incoherence in the unsafety system of Rust that I'd love to help fix.

> I agree that it is unclear the way that unsafe annotates two different sides of the safety contract. However, I am not as convinced that this change will make a big impact on ellucidating the distinction, because we are still using the unsafe keyword in both positions. I think the use of one keyword for both roles (which I think we aren't going to change) is the real stopper here, not the fact that in some cases the keyword plays multiple roles.

I feel like it still helps, because it means every `unsafe` is *one or the other*, whereas currently some `unsafe` actually play *both* roles.

**Comment by faern** [comment]
I think it boils down very much to your style of coding. I have found myself on the opposite side of the spectrum compared to @withoutboats a few times recently. When prototyping, I have actively avoided making functions that are unsafe to call `unsafe fn`s. Because it's like turning off all safety checks and stops the compile from helping me. When I prototype I want to properly see where I do unsafe stuff so I can reason about the invariants I have to uphold. So in order to have the compiler tell me when I do something unsafe I must avoid `unsafe fn`s, and then I try to remember to slap that `unsafe` on the function when prototyping and `//SAFETY:` documentation is done. And that feels risky. Both because I can forget to add the `unsafe` when I'm done, and after I have added it it becomes harder to recognize new unsafe operations in that function and to properly reason about and document them.

I love what this RFC is doing.

**Comment by nikomatsakis** [comment]
> I also proposed some ideas once to be able to mark struct fields as unsafe, which would make mutating them an unsafe operation. 

Side note, but @RalfJung I am still in favor of this change and would like to see it go through.

**Comment by Plecra** [comment]
It seems like a new RFC would need to be made to elaborate on the motivation for this change at the moment, but I think the use case presented by `ringbahn::Event` is particularly important for developers writing *safe* code.

While there's little reason to write an unsafe function with a safe body, the implementor of an unsafe trait method doesn't control the signature. If they want to keep to the safe set of Rust (`forbid(unsafe_code)`), they could write a perfectly safe implementation, maybe with runtime checks. But with the implicit `unsafe` block they're left with, it will still need to be manually verified.

There are a few nightly traits that use the same pattern, and it's helpful for 'optimisations' in things like bounds checks. I think forcing developers into `unsafe` Rust is a real shame when the language is otherwise so flexible (or "empowering" by the tagline)

**Comment by jonhoo** [comment]
> While there's little reason to write an unsafe function with a safe body

Interesting.. This does not match my experience very well — even within my `unsafe fn`s, _most_ of the code is safe, except for _some_ parts that are unsafe.

**Comment by Plecra** [comment]
Agreed! I meant to refer to `unsafe fn`s with *no* unsafe code in their bodies (and I strongly support proposals like unsafe fields for cases like `Vec::set_len`).

Since I was struggling to explain my reasoning, maybe this example could elaborate on it. With tools like `cargo-geiger` popping up, I think it's quite important that Rust considers this code "safe":

```rust
trait Collection<E> {
    /// # Safety
    /// `index` must be a valid index into this collection
    unsafe fn get_unchecked(&self, index: usize) -> E;
}
struct Bucket<T> {
    value: T,
}
impl<T: Copy> Collection<T> for Bucket<T> {
    unsafe fn get_unchecked(&self, _: usize) -> T {
        self.value
    }
}
```

Afaik, Rust doesn't provide any way to verify that this kind of code is safe atm, due to the implicit block this issue is addressing.

**Comment by SimonSapin** [comment]
Could a way to opt-in to this new behavior be stabilized without blocking on decisions of when or how to change the default?

**Comment by RalfJung** [comment]
AFAIK the lint does nothing when not enabled, so in principle, couldn't we just remove the feature gate?

**Comment by nikomatsakis** [comment]
Indeed, I think we could stabilize the lint with a default level of allow and separately consider whether to change the default.

**Comment by LeSeulArtichaut** [comment]
I can file a PR for this

**Comment by programmerjake** [comment]
Could the top comment be edited to indicate that `#![deny(unsafe_op_in_unsafe_fn)]` is stable?

**Comment by Allen-Webb** [comment]
> Could the top comment be edited to indicate that `#![deny(unsafe_op_in_unsafe_fn)]` is stable?

My understanding is this doesn't hit stable until 1.52.0 which isn't out yet. I tried to use it on on 1.51. 0 and got an error saying it was unstable.

**Comment by camelid** [comment]
> * Stabilization PR ([see instructions on rustc-dev-guide](https://rustc-dev-guide.rust-lang.org/stabilization_guide.html#stabilization-pr))

I think this step can be checked off now?

**Comment by Serentty** [comment]
Honestly, I worry that with this change, which makes unsafety much more verbose, many people will just be lazy and write an unsafe block inside their function (since the compiler wants them to), without bothering to mark the function as unsafe to call, even if it is. You can say “But they're not supposed to! That's now how any of this works!” But if you have to argue that, then you're dealing with a pit of failure instead of a pit of success, and so many Rust design decisions are explicitly about trying to make it easier to do things the right way than the wrong way.

**Comment by RalfJung** [comment]
@Serentty in which way do you think that this RFC makes things any worse there? I wouldn't call "huge `unsafe fn` without any fine-grained `unsafe` control inside it" to be a "pit of success" -- that's also a failure, just a different kind.

**Comment by Serentty** [comment]
For very long functions, that is a risk. For very short functions, I think the risk is in the other direction though.

**Comment by Allen-Webb** [comment]
@Serentty Do you have an ideas on how to resolve your concern? I would argue that laziness on the part of some developers should not be an excuse to prevent developers who are not lazy from being able to narrow the scope of the unsafe code in their code. It seems like to me that same argument could be applied to any Rust code---if people are lazy and don't want to solve the safety issues properly they may just put them in an unsafe block / unsafe function / etc. This problem is not really made worse by the RFC at all in my opinion, but instead may eventually make someone who is lazy go further out of their way to circumvent the safety features of the language.

**Comment by Serentty** [comment]
> I would argue that laziness on the part of some developers should not be an excuse to prevent developers who are not lazy from being able to narrow the scope of the unsafe code in their code.

I agree as long as it's a lint that can be turned on, and doesn't get cranked up to being on by default or worse, an error.

My concern is one about the amount of ceremony required in order to use unsafe code with this. Yes, I understand the conceptual motivation of separating unsafe to call from performing an unsafe operation. On the other hand, requiring too much ceremony can make Rust less ergonomic for use cases which by their nature require large amounts of unsafe code. For example, lately I've been developing for 6502-based platforms in Rust, which naturally involves poking at a lot of hardware registers. Yes, I am building safe abstractions for these things, but even then, the amount of code required to build the safe abstractions is significant, and I feel like something like this would just add line noise, without helping isolate safe from unsafe code in my functions, which are already very short, and do consist of nearly entirely unsafe code. For many of them, the ceremony required would be about as long as the function itself.

> Do you have an ideas on how to resolve your concern?

I do, actually. This is all well and good for people who want it and will turn it on. But before it's made the default, I would really want there to be some way to make it more concise than having to put your whole (often very short) function body in an unsafe block. One possible solution would be to have some sort of syntax for “this scope is all unsafe code, AND is unsafe to call” that can be applied to functions, and to separate that from the “unsafe to call” syntax. I think merely changing the meaning of the current syntax without providing an alternative for the old meaning is unfortunate.

**Comment by Serentty** [comment]
I should add that I _do_ sometimes write code which would benefit from this, by the way. But the use case that I described in the previous post where I have lots of very small functions which do things like poke hardware registers is _also_ a major way I use unsafe code, and I don't think it's really a bad or unidiomatic one, but rather a nice and modular one.

**Comment by phil-opp** [comment]
Even if the lint was changed to warn by default you could still `#[allow(...)]` it at the crate/module/impl level for crates that use a lot of unsafe.

**Comment by Serentty** [comment]
So then unsafe functions will be in a limbo land where they're still technically all one big unsafe block, but there's a lint against using them the way that they technically work. I think it makes more sense to just reserve some syntax for “function which is an unsafe block (and also unsafe to call)” separate from functions which are just unsafe to call. That way there's no ambiguity over what “unsafe” really means, and making it a hard error in a future edition wouldn't mandate extra tedium.

**Comment by Nemo157** [comment]
One solution I remember being mentioned (probably somewhere in the RFC thread) is expression functions to reduce the nesting (I'm not sure whether there's any issue open to track proposing these, IIRC they were something Centril planned on working on):

```rust
unsafe fn get_unchecked<T>(x: &[T], i: usize) -> &T = unsafe {
    x.get_unchecked(i)
};
```

**Comment by Serentty** [comment]
> One solution I remember being mentioned (probably somewhere in the RFC thread) is expression functions to reduce the nesting (I'm not sure whether there's any issue open to track proposing these, IIRC they were something Centril planned on working on):
> 
> ```rust
> unsafe fn get_unchecked<T>(x: &[T], i: usize) -> &T = unsafe {
>     x.get_unchecked(i)
> };
> ```

Wow, that's just a tiny bit more verbose than I would find ideal, but it's a huge improvement. I definitely think that a good syntax should be added before making the lint the default. If there is a concise syntax added, I don't have as much of a problem with any of it.

**Comment by runiq** [comment]
> ```rust
> unsafe fn get_unchecked<T>(x: &[T], i: usize) -> &T = unsafe {
>     x.get_unchecked(i)
> };
> ```

I don't see a big difference between that :point_up: and this :point_down::

```rust
unsafe fn get_unchecked<T>(x: &[T], i: usize) -> &T { unsafe {
    x.get_unchecked(i)
}};
```

**Comment by Robbepop** [comment]
> > ```rust
> > unsafe fn get_unchecked<T>(x: &[T], i: usize) -> &T = unsafe {
> >     x.get_unchecked(i)
> > };
> > ```
> 
> I don't see a big difference between that point_up and this point_down:
> 
> ```rust
> unsafe fn get_unchecked<T>(x: &[T], i: usize) -> &T { unsafe {
>     x.get_unchecked(i)
> }};
> ```
You do as soon as your favorite Rust formatter goes brrrrrrrr
```rust
unsafe fn get_unchecked<T>(x: &[T], i: usize) -> &T {
    unsafe {
        x.get_unchecked(i)
    }
}
```
Fewer braces reduce the cognitive complexity for readers for non trivial examples.
However, I personally am not in favor of the expression short hand syntax although I also do not feel strongly about it either.

**Comment by Serentty** [comment]
> > ```rust
> > unsafe fn get_unchecked<T>(x: &[T], i: usize) -> &T = unsafe {
> >     x.get_unchecked(i)
> > };
> > ```
> 
> I don't see a big difference between that ☝️ and this 👇:
> 
> ```rust
> unsafe fn get_unchecked<T>(x: &[T], i: usize) -> &T { unsafe {
>     x.get_unchecked(i)
> }};
> ```

I agree with @Robbepop that in concept, more braces tends to mean more cognitive overhead. When you conceptually have multiple scopes, that's justified, but in a case like this, the obvious way of doing it with the existing syntax turns what was one scope into one nested in another, I think it definitely increases line noise and mental parsing effort. It's not even nice and symmetric because it closes with double braces, but opens with braces with a keyword between them. I think “being pleasant to read” is a consideration that shouldn't be completely ignored.

**Comment by Allen-Webb** [comment]
Is there a parsing reason why we cannot have the following?
```
unsafe fn get_unchecked<T>(x: &[T], i: usize) -> &T unsafe {
    x.get_unchecked(i)
}
```

**Comment by mark-i-m** [comment]
@Allen-Webb In my opinion, the method header should contain only things that are relevant to the caller, not things that are relevant to the implementer. I think that's what makes the current state of `unsafe fn` so confusing to being with. The syntax you proposed better, but still IMHO sets a bad precedent...

**Comment by programmerjake** [comment]
How about:
```rust
unsafe fn get_unchecked<T>(x: &[T], i: usize) -> &T {
    #![unsafe]
    x.get_unchecked(i)
}
```

**Comment by mlindner** [comment]
> Second, so far two memory safety bugs have been discovered in ringbahn, and neither of them would have been prevented by this change. **In both cases, invariants were not being upheld by safe code (in both cases, interestingly, because of implicit destructors).** The bugs this change is designed to catch seem easy to identify and unlikely to occur: accidentally calling an unsafe function that you didn't mean to call. I am doubtful that this would meaningfully improve the quality of unsafe code or the experience of writing it.


Sorry if this is a bit of a necropost, but you should never be writing unsafe code that doesn't check all of its required invariants. If you rely on safe code to ensure certain invariants are held, then the unsafe code is incorrectly written.

**Comment by RalfJung** [comment]
@mlindner I do not agree with your statement in the generality that you are making it.

It is perfectly fine for unsafe code to rely on *known safe code* for its invariants. For example, if I write an entirely safe `sort` function, it is okay for unsafe code to call that function and rely on it to indeed correctly perform a "sort". Saying that unsafe code must not rely on safe code for its correctness is banning unsafe code from using a large part of the Rust ecosystem (including a large part of the standard library), which would not help anyone.

**Comment by ojeda** [comment]
> Should we ever make this deny-by-default or even a hard error, in a future edition?

In Rust for Linux we have been using `-Dunsafe_op_in_unsafe_fn` for a while and we think it is the right approach -- the concepts are orthogonal, after all. The ergonomics have not been a problem for us so far. For more context, we also require `// SAFETY: ` comments and we minimize & split `unsafe` blocks as needed.

(From a discussion at LPC today, @joshtriplett suggested to provide our feedback on this)

**Comment by joshtriplett** [comment]
Based on the various feedback we've received, I'm going to propose that we *should* enable this by default in the 2024 edition, assuming that we have a rustfix that adds unsafe blocks as appropriate (ideally fine-grained unsafe blocks).

@rust-lang/lang, shall we require `unsafe {}` for unsafe blocks even inside an `unsafe fn`, starting in the 2024 edition?

(If you have thoughts on deny-by-default vs warn-by-default, please chime in with those. I would propose deny-by-default, but I think either would be an improvement.)

@rfcbot merge

**Comment by Lokathor** [comment]
I believe that the general edition guidance is to begin warning this edition and then deny by default in edition+1

**Comment by Serentty** [comment]
So there is a plan to have it start denying by default? I really, really don’t think that that should be done until a more concise syntax is devised for “unsafe function whose body is also an unsafe block”. That is such an incredibly common use case, and having to double up on braces feels like it adds to line noise, and makes unsafe Rust even uglier than it currently is. Distinguishing between “unsafe to call” and “unsafe implementation” is good, yes, but I think there needs to be a more eloquent way to do it.

**Comment by Lokathor** [comment]
Just put an `allow` on your crate. The entire lint is very obviously insufficient for stuff like data structures until we *at least* have something like unsafe fields too, so until then many crates will just want to turn it off and keep using the previous code style.

**Comment by Serentty** [comment]
If it’s still insufficient, why would it be on by default? It makes more sense to work those things out before turning it on.

**Comment by Lokathor** [comment]
There's many cases where it *does* help make the code more clear. Particularly if the crate is the kind of thing where there's unsafe free functions that do work while calling other unsafe subroutines. In that case, particularly if the functions are medium to large, it helps to highlight the unsafe inner calls.

**Comment by scottmcm** [comment]
@rfcbot concern define-required

I don't think we should FCP this without a specific plan for exactly what we're deciding here.  It says "require", but that could be everything from lint-warn to hard-error.

From what I can tell from <https://github.com/rust-lang/rust/search?q=unsafe_op_in_unsafe_fn+path%3Asrc%2Ftest>, it seems like there's not yet a machine-applicable fix for the lint either.  I think knowing what a post-edition-update `fix` would do to code is important for understanding the impact of suggesting this as default.

**Comment by Lokathor** [comment]
there's two machine updates that seem reasonable:
* wrap the entire function body in an unsafe block.
* just mark the crate as allowing the lint.

**Comment by faern** [comment]
I'd like to add:

- wrap each unsafe statement in separate unsafe blocks.

That's usually how I try to code. So there is a logical place to put the safety documentation for each unsafe statement.

**Comment by Lokathor** [comment]
But making several small unsafe spans doesn't document the code. Only a human can do that. So I think the auto-fix should lean towards making it very clear that there was just a blanket fix applied, not a careful inspection.

**Comment by pnkfelix** [comment]
> there's two machine updates that seem reasonable:
> 
>     * wrap the entire function body in an unsafe block.
> 
>     * just mark the crate as allowing the lint.

I'm in favor of a rustfix that takes the first listed approach ("wrap the entire function body in an unsafe block"), because we cannot infer whether the crate author is someone who uses `unsafe { ... }` as something to highlight unsafe operations, or if the crate author is someone who uses `unsafe { ... }` to demarcate the lexical scope during which runtime safety invariants are broken (and must be re-established before the end of the scope).

To handle the latter case, wrapping the entire function body in an unsafe block seems like the only reasonable thing that is nearly consistent with the whole proposal here (and even *there*, there is the possibility of someone who has in fact written an `unsafe fn` that *doesn't* re-establish all of the runtime safety invariants before the `unsafe fn` itself returns).

**Comment by joshtriplett** [comment]
We discussed this in today's @rust-lang/lang meeting. Summary:

- We don't want to make this warn-by-default in 2021 or earlier, because we were still in 2021 emitting a warning telling people to *not* do this.
- We don't want to make this a hard error in 2024. We're skeptical about doing so in any future edition either.
- We're still talking about whether it should be warn-by-default or deny-by-default in 2024.
- We definitely want a rustfix in order to make this change.
- We're still talking about how fine-grained the rustfix should be, but it sounds like it may be *most* feasible to make it a single coarse block, either around the whole function or around all the unsafe code in the function.

**Comment by joshtriplett** [comment]
(My own intent with the FCP was to confirm "we're definitely doing either warn or deny, so it's worth working on the rustfix and related aspects".)

**Comment by Lokathor** [comment]
[2502-epochs](https://github.com/rust-lang/rfcs/blob/master/text/2052-epochs.md)

> code that compiles without warnings on the previous edition (under the latest compiler release) will compile without errors on the next edition (modulo the [usual caveats](https://github.com/rust-lang/rfcs/blob/master/text/1122-language-semver.md) about type inference changes and so on).

If it's not warn by default at any point in 2021 edition, the 2024 should stick to only warn by default, not deny by default.

**Comment by Lokathor** [comment]
**Having thought about the matter more:** I think for the warning level, it does not matter what we did in past *editions*, it just matters what we did in past *compiler versions*. It has been 2 years since the RFC was accepted and Rust began to move away from the previous style to the new style.

If we're saying that this lint is an appropriate guidance in edition 2021 (which we are), and there's nothing edition 2021 changed about unsafe code compared to edition 2018 (which there was not), then it must be appropriate guidance in edition 2018 as well. And the exact same holds going from edition 2018 back into edition 2015. So if it's ever a warning, this should be a warning in *all* editions leading up to the deny-by-default edition.

**Comment by LeSeulArtichaut** [comment]
Opened https://github.com/rust-lang/rust/pull/99827 as a starting point, which simply adds an `unsafe` block wrapping the entire function with no regards to proper style.

Having finer-grained unsafe blocks sounds possible to me, though we'd have to be careful around field expressions, places, etc.

**Comment by QuineDot** [comment]
> We don't want to make this warn-by-default in 2021 or earlier, because we were still in 2021 emitting a warning telling people to _not_ do this.

Isn't the first step to simply stop warning on `unsafe` in `unsafe`, or at a minimum those that contain unsafe operations?

**Comment by Lokathor** [comment]
Ah, Yes. My previous long comment was based on the (apparently very wrong!) notion that we'd stopped doing this ages ago. I thought that had already happened.

We should stop doing that *immediately* (in all editions), wait some number of versions (6? however many), and then warn in the opposite direction (again, in all directions). Then a future edition would be well placed to potentially deny-by-default.

**Comment by LeSeulArtichaut** [comment]
> Isn't the first step to simply stop warning on `unsafe` in `unsafe`, or at a minimum those that contain unsafe operations?

It's already the case if the `unsafe_op_in_unsafe_fn` lint is `warn` or greater. This doesn't yield any warnings:

```rs
#![warn(unsafe_op_in_unsafe_fn)]

unsafe fn unsf() {}

pub unsafe fn foo() {
    unsafe { unsf() }
}
```

In other words, we consider that an unsafe block is "needed" if the absence of said unsafe block would lead to at least a warning.

**Comment by RalfJung** [comment]
Yeah that was the consensus when the RFC got passed.

If there is consensus for that it should be easy to always remove the warning.

**Comment by joshtriplett** [comment]
Yikes. @QuineDot Good catch.

We should disable `unused_unsafe` ASAP.

And given that, we may have to make this warn-by-default in 2024 rather than deny-by-default. :(

**Comment by scottmcm** [comment]
With https://github.com/rust-lang/rust/issues/71668#issuecomment-1189396860 updated to specify that this is "warn in 2024",

@rfcbot resolve define-required

**Comment by joshtriplett** [comment]
Without waiting for the FCP for this, could someone submit a PR changing `unused_unsafe` to allow?

**Comment by Lokathor** [comment]
> And given that, we may have to make this warn-by-default in 2024 rather than deny-by-default. :(

I still don't think so. This is still a version issue, not an edition issue.

**Comment by RalfJung** [comment]
> could someone submit a PR changing unused_unsafe to allow?

Done: https://github.com/rust-lang/rust/pull/100081

**Comment by leddoo** [comment]
hey, i just found the `unsafe_block_in_unsafe_fn` rfc.

here's an alternative that *does not* require changing the current behavior *and* improves the overall *granularity* of the `unsafe` mechanism: a `safe` block.

basically, introducing a `safe` block switches back to normal safe rust.
```rust
fn foo() {
    // no unsafe stuff here!
    unsafe {
        // can do unsafe stuff.
        safe {
            // can no longer do unsafe stuff.
            unsafe { /* you get the idea */ }
        }
    }
}
```

combining this idea with RalfJung's idea, we get
```rust
unsafe fn unsafe_to_call_but_body_is_safe() safe {
    // safe by default
    unsafe {
        // opt into unsafety.
    }
}
```

**Comment by leddoo** [comment]
even if the `unsafe_block_in_unsafe_fn` rfc gets (or is already) accepted, this would be a nice thing to have.

**Comment by digama0** [comment]
I believe this already exists (or existed?) in the rust internals, it is needed to do unsafety checking in macros when you have a match-like construct where the match itself needs to be unsafe but the match arms are not, so you can't use strict block scoping. There was some kind of stack semantics where you can cancel n nested `unsafe {` blocks with n nested `safe {` blocks.

**Comment by Subsentient** [comment]
I'm trying to understand this decision process. I am very concerned and annoyed.

So if I want my code to be future-proof, I need to start adding unsafe{} all over all my unsafe functions? Every single one of them that does actually unsafe stuff?

Like I often did already in the beginning but was repeatedly told not to by rustc, so stopped?

Is that correct?

Don't tell me to use the lint, you've already made explicit that it will eventually be a hard error.

You do know how much code is already using "unsafe fn" without a second unsafe{}? Probably most of it. The amount of code that will break if you deny-by-default?

You intend to make already deeply ugly, cumbersome unsafe code even more painful to use?

I have to tell you, if I was a total Rust newbie but experienced with other systems languages, seeing this kind of planning by core development like on this issue would *strongly* deter me from investing further in Rust, because it would make me seriously question the judgement of maintainers and worry they'll just constantly break more code, mine and dependencies, in the near future.

I am trying to start a Wayland compositor project in Rust right now. I was very encouraged, this just was a gut-punch to my motivation.

This change will probably require modification to every single Rust package that uses unsafe in the ecosystem. It was tolerable before rustc was actually stable, the language was still "evolving". Not anymore. You will find this excruciatingly painful when abandoned but widely used packages on crates.io break many, many builds. Is it also not humiliating for the core team to have to totally reverse directions on this warning?

I'm sorry if my tone is a bit aggressive, I'm quite perturbed with this and frankly shocked to find this RFC tonight.

**Comment by NobodyXu** [comment]
I definitely can understand why this change is made: This RFC basically wants to separate `unsafe` function interface from `unsafe` function implementation, stating that an `unsafe` function interface does not mean that the implementation is necessary `unsafe`.

But the breakage mentioned by @Subsentient , while I am not mad or angry I do think that's a problem.
IMO this kind of changes requires an edition to made and even so it might cause a lot of crates to stay on old edition as the breakage takes quite long to fix.

I suggest to run the crater on all crates at crates.io with this changes to `rustc` applied to get a general idea of how many crates would be broken by this.

My guess is probably majority of crates that uses `unsafe fn ...` will be broken by this and ones that are not broken are the minorities.

I agree with @withoutboats that this is probably not very helpful at all and many are just going to wrap the whole function in `unsafe {}`.

I suppose this is better as a lint instead of a hard error for everybody, thus I am against of merging this PR in its current state.

**Comment by Subsentient** [comment]
I wouldn't like it, but I'd be fine if this was a clippy-only thing. The language should always continue to allow unsafe fn{}s to do unsafe things, no warnings, but I'd be fine if it was in one of the higher warning levels of clippy. If you do anything to deny compilation of this code, people will be angry. This is a stylistic choice. I understand it makes things slightly safer, I agree, but my main assertion is that this cannot be allowed any way outside of clippy. As @NobodyXu suggested, I would _strongly_ support testing crates.io to see how many crates are affected, and I think at bare minimum, that should be done before any denial-related functionality hits rustc.

> I agree with @withoutboats that this is probably not very helpful at all and many are just going to wrap the whole function in unsafe {} 

As do I. That is exactly what I will do, and many including myself will resent the devs for the forced extra braces and indentation.

**Comment by Subsentient** [comment]
> I definitely can understand why this change is made: This RFC basically wants to separate `unsafe` function interface from `unsafe` function implementation, stating that an `unsafe` function interface does not mean that the implementation is necessary `unsafe`.
> 
> But the breakage mentioned by @Subsentient , while I am not mad or angry I do think that's a problem. IMO this kind of changes requires an edition to made and even so it might cause a lot of crates to stay on old edition as the breakage takes quite long to fix.
> 
> I suggest to run the crater on all crates at crates.io with this changes to `rustc` applied to get a general idea of how many crates would be broken by this.
> 
> My guess is probably majority of crates that uses `unsafe fn ...` will be broken by this and ones that are not broken are the minorities.
> 
> I agree with @withoutboats that this is probably not very helpful at all and many are just going to wrap the whole function in `unsafe {}`.
> 
> I suppose this is better as a lint instead of a hard error for everybody, thus I am against of merging this PR in its current state.

Sadly it appears it's too late, this RFC was already approved. All we can do is sit back and wait for the mushroom cloud to be visible. Once this hits a rustc people use, you will need to be careful using cargo and rustup, even if your own code is already updated. Keep your old Rust edition on hand, but that might choke too, from new updates to all your dependencies, e.g. 10,000 warnings from all packages.

I guess we just wait for the sirens.

**Comment by Nilstrieb** [comment]
There appears to be some confusion going on. Cargo compiles all dependencies with the `--cap-lints` flag, meaning lints in dependencies are never shown to the user or never break a build if the crate `deny`s or `forbids`s them. So when this gets warn by default, a user will _never_ see this from a dependency (and it also won't impact performance or anything like that if you're concerned about that).
Also, if it's made warn-by-default only from edition 2024, crates on the older editions don't have to care anyways.
When you want to update your crate to the next edition, you will have to insert a bunch of unsafe blocks if you use unsafe operations in unsafe functions. This sounds bad, but just like with all editions, `cargo fix --edition` should automatically do the upgrade for you, so you just have to run that to be updated, no manual changes needed (except maybe a quick skim over it to make sure nothing broke).
So there will definitely not be any "mushroom cloud" or "sirens", it will all be fine. Worst case you don't like the style change from this to your codebase. I can't say anything on that, the lang team decided that the new style is better (which I agree with). I'm sure you will be able to live with it.

**Comment by Subsentient** [comment]
@Nilstrieb Are you quite sure of that? Or perhaps I'm not understanding correctly?

![image](https://user-images.githubusercontent.com/1091969/211154463-f600a6c3-b348-4edc-9dfd-b5bce6b97d6c.png)

**Comment by Subsentient** [comment]
Also, in case you meant #![deny()] had to be on: 
![image](https://user-images.githubusercontent.com/1091969/211154602-4300d49f-4e54-43ee-a9be-74690c6307c2.png)

**Comment by Nilstrieb** [comment]
`cap-lints` is only on for third-party dependencies (like those that you get from crates.io or `cargo vendor`), not path dependencies.

**Comment by Subsentient** [comment]
@Nilstrieb Ahh, I see, no doubt not documented anywhere, only known by rustc/cargo devs. That fits about what I expected.

I still have great doubt that this change is going to go down smooth, and I'm not happy with it, but I guess some people are.

Alright, I'll just look away and wait to see if there's a fission flash or not when this rolls out. All that said, if there are cargo issues, I will absolutely be back to remind you of this conversation.

**Comment by RalfJung** [comment]
There are no plans to break any existing code here, so please stop the hyperbole @Subsentient. Also your tone is non-constructive and unhelpful. If you think something is missing from the cargo docs, please file an issue.

**Comment by Ciel-MC** [comment]
My opinions probably don't matter but:
I think this should be warn by default now and have
```rust
fn foo() unsafe {} // or = unsafe
```
before it's made into deny/error in 2024.

**Comment by withoutboats** [comment]
(NOT A CONTRIBUTION)

My opinion today is that turning this into a warn by default lint at some point is a good choice. As a warn by default lint, the concern I expressed about prototyping vanishes - you can just ignore the lint while experimenting the same way you ignore unused variable and dead code lints. In fact, I think that with this as a warn-by-default lint, you get essentially the same experience I talked about with a "requiring annotations on every unsafe operation" lint, for less effort.

I'm against a deny by default lint for the same reason: in order to make progress while prototyping, users would have to turn it off. But it's actually helpful for users to see this signal while they're working, and I think users are more likely to allow it than turn it to a warning if its deny by default.

The only concern I have still is the larger concern about the UX of unsafe and the fact that this isn't coupled with other improvements to the language, standard library, and documentation that makes it easier to understand what to do when writing unsafe code. Alone, I worry for many users this will just come across as ineffectual busywork, because Rust hasn't done enough to enable them to really understand what the invariants their unsafe code is maintaining *are*. It's just the reality that a lot of unsafe code is being written without the kind of clear minded correctness checking that project would like; the question is how do we get that to a better place? I feel like the Rust project's solutions have been all stick and no carrot.

For example, I'm really in favor of changes like this one Ralf mentioned:

> I also proposed some ideas once to be able to mark struct fields as unsafe, which would make mutating them an unsafe operation. Then unsafe tracking would help even with maintaining module invariants. (This could also affect auto-trait derivation, but that is yet another discussion.) That is an orthogonal feature but could be part of a "vision" for making unsafe more helpful. Or maybe it's not actually helpful, I am not sure.

However, work toward these kinds of changes is intermittent and slow, so I'm not saying that I think making this lint a warning should be blocked on figuring out a larger unsafe ergonomics story, which will occur on an unknown timeline. But I am concerned that turning it on without that may just do damage to the intended effect: e.g. a lot of orgs may choose to have a policy of permanently silencing it because writing unsafe code properly is too much work.

**Comment by Nemo157** [comment]
There's a PR open to make it warn-by-default for edition 2024, following the [FCP from last year](https://github.com/rust-lang/rust/issues/71668#issuecomment-1189396860), just blocked on getting someone to approve it https://github.com/rust-lang/rust/pull/112038

**Comment by awnion** [comment]
(NOT A CONTRIBUTION)

TLDR: I'm pro lint #![warn. I'm against a #![deny by default.

Nobody mentioned that `unsafe` syntax behavior is aligned with `async` syntax behavior _Id est_:

```rust
unsafe fn u() {
    // whole block is unsafe {...}
}

async fn a() {
    // whole block is async {...}
}
```

Lint warning I think would be good enough from just a pedagogical perspective. Deny by default would be just annoying at best.

**Comment by RalfJung** [comment]
`async` and `unsafe` are keywords of very different nature, so the fact that they are aligned is actually *bad*. `async` is a form of effect; effects need to be propagated outwards. `unsafe` is not an effect, it is a proof obligation; it actually has two different meanings depending on where it is used. See the [keyword docs](https://doc.rust-lang.org/nightly/std/keyword.unsafe.html#the-different-meanings-of-unsafe) for more discussion of this topic.

**Comment by awnion** [comment]
> `async` and `unsafe` are keywords of very different nature, so the fact that they are aligned is actually _bad_. `async` is a form of effect; effects need to be propagated outwards. `unsafe` is not an effect, it is a proof obligation; it actually has two different meanings depending on where it is used. See the [keyword reference](https://doc.rust-lang.org/nightly/reference/unsafe-keyword.html) for more discussion of this topic.

I agree. I just felt like would be nice to point out this similarity.

Also if you'd ask me I'm against both `async fn` syntax and `unsafe fn`.
I'd be good with `fn ... -> impl Future<...>` only. The same with `unsafe`. But we have what we have...

**Comment by digama0** [comment]
> I'd be good with `fn ... -> impl Future<...>` only. The same with `unsafe`. 

What does that mean? As Ralf said, they are very different features so I don't know how to analogize that to get "the same with `unsafe`" - `fn ... -> impl Unsafe<...>` makes no sense.

**Comment by awnion** [comment]
> > I'd be good with `fn ... -> impl Future<...>` only. The same with `unsafe`.
> 
> What does that mean? As Ralf said, they are very different features so I don't know how to analogize that to get "the same with `unsafe`" - `fn ... -> impl Unsafe<...>` makes no sense.

You're right. Just wanted to share a perspective on how someone might see it.

**Comment by pigeonhands** [comment]
If `unsafe` was changed so that the body of the function no longer has a `unsafe` block around it, a macro/annotation could be used to return the old behavior to solve the indent issue.

So for example, these two functions would be identical
```rust
#[unsafe_body]
unsafe fn foo() { 
    my_unsafe_function()
}
```

```rust
unsafe fn foo() { 
  unsafe{
      my_unsafe_function()
  }
}
```

**Comment by traviscross** [comment]
Note that landing this in Rust 2024 probably depends on resolving #119823.

**Comment by traviscross** [comment]
Note that #119823 is now resolved, clearing the way forward for this.

**Comment by traviscross** [comment]
@rustbot labels +I-lang-nominated

Given the resolution of #119823, let's discuss what else (if anything) needs to happen to land this in Rust 2024.

**Comment by traviscross** [comment]
@rustbot labels -I-lang-nominated

We discussed this today in triage and agreed that we would like to make this a warn-by-default lint in Rust 2024 with the intent to later make this a warn-by-default lint across all editions.

We will be proposing FCP merge to this effect.

**Comment by tmandry** [comment]
I've opened #120535 for making this lint warn-by-default in the 2024 edition and proposed FCP merge.

**Comment by traviscross** [comment]
We'll track the aspect of this related to Rust 2024 separately here:

- https://github.com/rust-lang/rust/issues/123916

**Comment by Serentty** [comment]
This being warn-by-default in 2024 when there is still no more-concise syntax for “unsafe function with unsafe body” is unfortunate, in my opinion. When I pointed out in the past that this adds an enormous amount of line noise to functions that peek and poke at memory as is typical when writing bare-metal code, a common argument was that this could be reduced by introducing a more concise syntax. However, by switching this to warn-by-default *without* first providing such a syntax, people are encouraged to switch to this much more verbose syntax right away, through compiler warnings. If an easier way ever is added later, it will mean switching over codebases *twice*. Plus, once the safety considerations are solved, there is much less incentive to solve the syntactic ones, which is why I think it makes sense to solve them at the same time.

As it stands, this means that such functions have to have two sets of braces.  To the extent that it might be possible to simply the formatting of this visually with something like:

```rust
unsafe fn function() { unsafe {
    ...
}}
```

... this is not easily handled by `rustfmt`, making it hard to actually maintain such a style consistently.

Of course, there are always arguments such as “a more concise syntax could be added later” and “`rustfmt` could be updated to support the style above,” but all of this ignores that none of these things are available *now* while the change in syntax is being pushed on Rust programmers *now*. Sure, you can disable the lint, but that fractures the ecosystem stylistically between those that adopt the current new style being pushed, and those who wait for something more concise.

While it is too late for Rust 2024, I *strongly* believe that either a more concise syntax should be in the next edition, or this should not be warn-by-default.

**Comment by Lokathor** [comment]
just turn off the warning at the module or crate level and get it over with.

**Comment by Serentty** [comment]
> just turn off the warning at the module or crate level and get it over with.

I mentioned exactly that point in my comment, but okay.

**Comment by Lokathor** [comment]
I know you said that, and I'm sorry for giving a short reply that's in retrospect dismissive.

I'll give a slightly longer explanation of my point: when this was discussed ages ago in the T-lang triage meetings (which i was observing for at least one of the ones discussing this RFC, definitely not all of them), "just turn off the lint" was the *actual* proposal for how people would deal with this warning. That was really the acceptable enough answer for the people in the room at the time. One can turn off the warning at the function level (which is a line of code but which avoids rightwards drift), or if you've got enough functions in a module doing the same thing you can of course turn it off at the module level. Warnings are just warnings (and not hard errors) *so that* they can be turned off in the appropriate situation. It's totally reasonable in terms of code organization to have "the module where we keep the unsafe toolkit", and try to keep the more dangerous stuff in a limited number of files. Encouraging that kind of organization is basically a good thing.

**Comment by Serentty** [comment]
> I know you said that, and I'm sorry for giving a short reply that's in retrospect dismissive.
> 
> I'll give a slightly longer explanation of my point: when this was discussed ages ago in the T-lang triage meetings (which i was observing for at least one of the ones discussing this RFC, definitely not all of them), "just turn off the lint" was the *actual* proposal for how people would deal with this warning. That was really the acceptable enough answer for the people in the room at the time. One can turn off the warning at the function level (which is a line of code but which avoids rightwards drift), or if you've got enough functions in a module doing the same thing you can of course turn it off at the module level. Warnings are just warnings (and not hard errors) *so that* they can be turned off in the appropriate situation. It's totally reasonable in terms of code organization to have "the module where we keep the unsafe toolkit", and try to keep the more dangerous stuff in a limited number of files. Encouraging that kind of organization is basically a good thing.

I guess the difference in mindset here is that I tend to interpret warnings from the compiler as fundamentally telling me that something is bad style that should be avoided. But exactly what warnings _mean_ is a deep question and I am sure that there are many valid interpretations. 

Anyway, I find this approach a bit strange, but I do appreciate you taking the time to answer. I am at least lukewarmly satisfied if “configure this at the function or module level” is the intended long-term solution, and the intent is not to turn this into an error in the future.

**Comment by RalfJung** [comment]
I'm afraid your intuition is not the one intended by Rust, which leads your mental model to clash with the way Rust language designers are thinking. `unsafe` is not an effect / a monad, it is an entirely syntactic marker indicating the presence of preconditions. It has no bearing whatsoever on the runtime behavior of a program.

Please consult https://doc.rust-lang.org/nightly/std/keyword.unsafe.html for some further reading.

**Comment by 9thyear2** [comment]
> > I would note that two examples mean two different things.
> 
> Yes, but not from a compiler feedback perspective.
> 
> Previously: `unsafe fn { op }`: mostly opted for default, unsafe propagates `fn { unsafe op }`: inline unsafe blocks are slightly less convenient, therefore force the programmer to think about whether it's actually safe to call op `unsafe fn { unsafe op }`: almost noone used this
> 
> Now: `fn { unsafe op }`: what most people will opt for when writing **new** code. `unsafe fn { unsafe op }`: almost noone will use this for new code. Only for the lint this change introduced in **existing** code. Maybe not even that.
> 
> > In my opinion what you want is the following (which I think the RFC is suggesting, but perhaps the lint doesn't make clear enough).
> 
> No, that's clear enough, as should be to anyone who bothers to look up the docs. My point is exactly that this pattern (which should be the "correct" one) is not **enforced** and unclear why it would be needed just based on the compiler feedback. It is also inconvenient, therefore people are less likely to use it.

i tend to agree, i just came out of learning rust and i ended up asking myself when should i use an:

```
unsafe fn something () {
    unsafe {
        do_something()
    }
}
```

over

```
fn something () {
    unsafe {
        do_something()
    }
}
```

as there is no constraint to stop me using the second pattern over the first (i mean i defined one for myself (a personal rule (anytime i use raw pointers, it needs to be in an unsafe fn)), but to my knowledge that's not upheld by the compiler

your right and people don't want to admit it, people will begin to prefer the second pattern over the first one (not just devs, but those who are just getting started with rust too (who will just want to make the warnings and error disappear))

whats to stop people from defining safe functions that use a bunch of unsafe blocks (and making a "safe" function that's actually turns out to be unsafe)? while that was possible before. Requiring unsafe blocks in unsafe functions will make people prefer to just do that, as there would be almost no distinction when writing the code between an unsafe function and a regular function, as both would require unsafe blocks

what is needed for this to make sense is some forced use of unsafe functions, but if that doesn't accompany this. than this alone doesn't make sense

people are just like lightning... they take the path of least resistance. In this case that would be not typing the extra 7 characters `unsafe ` right before `fn`

from a human psychology standpoint, combined with rusts goals... this doesn't make any sense alone

**Comment by SimonSapin** [comment]
> as there is no constraint to stop me using the second pattern over the first […] but to my knowledge that's not upheld by the compiler

`unsafe` is all about things the compiler cannot entirely check on its own. That’s why it’s called unsafe. I’m sorry but when using `unsafe` at all there is no substitute for learning the meaning of these two patterns and using them accordingly. Namely: an `unsafe fn` is unsafe to call: the caller is responsible for maintaining some invariant (which should be documented). `fn` is safe to call with any parameter it accepts, and if an `unsafe {…}` block is used the function body itself is responsible to uphold the relevant invariants (ideally also with a comment that explains how).

**Comment by 9thyear2** [comment]
> compiler cannot entirely check on its own

i wasn't referencing the compiler checking for safety, but what is being called within the code (eg. not allowing the use of raw pointers outside a specific pattern as i described in my personal rule)

the problem i was pointing out... is previously there were 2 ways of writing unsafe code:

```
unsafe fn something () {
    do_something()
}
```
and
```
fn something () {
    unsafe {
        do_something()
    }
}
```

this would cause a dev to stop and think what is more appropriate because of the differences between how the code could be wrote

but now you did the meme ("they're the same picture"):
```
fn something () {      // <--- unsafe before fn or not... doesn't matter
    unsafe {
        do_something()
    }
}
```

regardless whether its safe so not, that stop and pause is now removed, as there is no difference in the code between writing a safe or unsafe function

you might define one in the docs, but the docs aren't right in your face as your writing the code. and most tutorials that people watch use an older version of rust where this wasn't a thing

so as stated before... when presented with this the majority of people will act just like lightning, and take the path of least resistance (or use this all the time):

```
fn something () {
    unsafe {
        do_something()
    }
}
```

and that personal rule i mentioned... was for me to introduce a differentiator back between the 2 styles (so i have a stop and pause)

you should look at doing the same

**Comment by ojeda** [comment]
Just to clarify: there are safe functions that use unsafe code inside them, and that is fine and expected -- your raw pointer personal rule sounds like it will make you mark functions as unsafe that should have been safe.

One could argue that `safe fn` should have been introduced as a way to opt-out, but the current way is simpler and more regular, in the sense that each of the positions of the keyword does something orthogonal to the other.

I would suggest using Zulip to discuss this though, since there are probably quite a lot of subscribers in this issue and any change around this would require a lot of discussion.

Thanks!

**Comment by briansmith** [comment]
> ```
> fn something () {      // <--- unsafe before fn or not... doesn't matter
> ```

It does matter. "`unsafe fn`" means that *users* of your function need to uphold safety invariants that the Rust language doesn't enforce, when calling your function. `fn` without `unsafe` means that there is nothing special that the caller needs to do. This is why every combination, all of the following, is needed:

```
unsafe fn f() { unsafe { ... } }`
       fn f() { unsafe { ... } }
       fn f() { /*no unsafe*/ { ... } }
unsafe fn f() { /*no unsafe*/ { ... } }
```

I guess writing the documentation that clarifies this, exactly to help people who find it confusing, is the main reason why this issue is still open (bullet point "Adjust documentation to describe the (somewhat unusual) effect of the lint". Not really any reason to stay subscribed to this for anybody not interested in that.

**Comment by 9thyear2** [comment]
> It does matter. "unsafe fn" means that users of your function need to uphold safety invariants that the Rust language doesn't enforce, when calling your function.

why do you guys insist on misinterpreting what i was saying on purpose

wasn't referring what its purpose was, it was to simplify it so i didn't have to put 2 code examples with a 7 character difference

it almost feels like what im saying is getting taken out of context to reaffirm this decision 

i was trying to be a little more professional and not use an actual meme, but since you insist on that here you go then:

<img width="340" height="390" alt="Image" src="https://github.com/user-attachments/assets/0d161151-b34d-4811-9501-88bddfbcc05e" />
-

all i was saying is a subtle difference in how you write the code will cause someone to pause and think about how best to proceed (yes my personal rule isn't perfect, but neither is expecting everyone to have the rust docs pulled up on a second screen to make sure the they are using each pattern correctly (or memorize the docs))

what i was saying was introduce a subtle difference between the 2 styles again (my personal rule was an example (hence the word "personal")), but at least is causes me to take a pause and think what i am doing
